selfdrive/locationd/torqued.py

```diff
@@ -0,0 +1,230 @@
+        steer_override = np.interp(np.arange(t - MIN_ENGAGE_BUFFER, t, DT_MDL), np.array(self.raw_points['carState_t']) + self.lag, self.raw_points['steer_override']).astype(bool)
+        vego = np.interp(t, np.array(self.raw_points['carState_t']) + self.lag, self.raw_points['vego'])
+        steer = np.interp(t, np.array(self.raw_points['carControl_t']) + self.lag, self.raw_points['steer_torque'])
+        lateral_acc = (vego * yaw_rate) - (np.sin(roll) * ACCELERATION_DUE_TO_GRAVITY)
+        if np.all(active) and (not np.any(steer_override)) and (vego > MIN_VEL) and (abs(steer) > STEER_MIN_THRESHOLD) and (abs(lateral_acc) <= LAT_ACC_THRESHOLD):
+          self.filtered_points.add_point(float(steer), float(lateral_acc))
+
+
+def main(sm=None, pm=None):
+  config_realtime_process([0, 1, 2, 3], 5)
+
+  if sm is None:
+    sm = messaging.SubMaster(['carControl', 'carState', 'liveLocationKalman'], poll=['liveLocationKalman'])
+
+  if pm is None:
+    pm = messaging.PubMaster(['liveTorqueParameters'])
+
+  params_reader = Params()
+  CP = car.CarParams.from_bytes(params_reader.get("CarParams", block=True))
+  torque_params = params_reader.get("LiveTorqueParameters")
+  estimator = TorqueEstimator(CP, torque_params)
+
+  while True:
+    sm.update()
+    if sm.all_checks():
+      for which in sm.updated.keys():
+        if sm.updated[which]:
+          t = sm.logMonoTime[which] * 1e-9
+          estimator.handle_log(t, which, sm[which])
+
+    if sm.updated['liveLocationKalman']:
+      # print(sm.frame, [len(v) for v in estimator.filtered_points.buckets.values()])
+      msg = messaging.new_message('liveTorqueParameters')
+      msg.valid = sm.all_checks()
+      liveTorqueParameters = msg.liveTorqueParameters
+
+      if estimator.filtered_points.is_valid():
+        try:
+          slope, offset, friction_coeff, steer_data_distribution, lat_acc_data_distribution = estimator.estimate_params()
+          # print(slope, offset, friction_coeff)
+        except Exception as e:
+          # print(e)
+          slope = offset = friction_coeff = None
+          cloudlog.exception(f"Error computing live torque params: {e}")
+
+        if estimator.is_sane(slope, offset, friction_coeff, steer_data_distribution, lat_acc_data_distribution):
+          liveTorqueParameters.liveValid = True
+          liveTorqueParameters.slopeRaw = float(slope)
+          liveTorqueParameters.offsetRaw = float(offset)
+          liveTorqueParameters.frictionCoefficientRaw = float(friction_coeff)
+          estimator.update_params({'slope': slope, 'offset': offset, 'frictionCoefficient': friction_coeff})
+        else:
+          cloudlog.exception("live torque params are numerically unstable")
+          liveTorqueParameters.liveValid = False
+          # estimator.reset()
+      else:
+        liveTorqueParameters.liveValid = False
+      liveTorqueParameters.slopeFiltered = float(estimator.filtered_params['slope'].x)
```

Comment:
TorqueEstimator should probably build this message, like calibrationd does https://github.com/commaai/openpilot/blob/1632c2aa62a2501783b7c4b3aa5b0b7dae7d00f3/selfdrive/locationd/calibrationd.py#L179